app/models/group.rb

```diff
@@ -1,73 +1,77 @@
 # frozen_string_literal: true
 
 class Group < ApplicationRecord
-  groupify :group, members: %i[users tasks], default_members: :users
+  has_many :group_memberships, dependent: :destroy
+  has_many :users, through: :group_memberships
+
+  has_many :group_tasks, dependent: :destroy
+  has_many :tasks, through: :group_tasks
+
   validates :name, presence: true
   validate :admin_in_group
 
-  def self.create_with_admin(params, user)
-    group = Group.new(params)
-    return group unless user
+  def add(user, role: :confirmed_member)
+    GroupMembership.new(group: self, user:, role:).save!
+  end
 
-    ActiveRecord::Base.transaction do
-      group.save(validate: false)
-      group.make_admin(user)
-      raise ActiveRecord::Rollback unless group.valid?
-    end
-    group
+  def group_membership_for(user)
+    group_memberships.where(user:).first
   end
 
   def admin?(user)
-    user.in_group?(self, as: 'admin')
+    admins.include? user
   end
 
   def confirmed_member?(user)
-    user.in_group?(self) unless user.in_group?(self, as: 'pending')
+    confirmed_members.include? user
   end
 
   def member?(user)
-    user.in_group?(self, as: 'member')
+    members.include? user
+  end
+
+  def applicant?(user)
+    applicants.include? user
   end
 
   def make_admin(user)
-    users.delete(user)
-    add(user, as: 'admin')
+    return false unless confirmed_member?(user)
+
+    group_membership_for(user)&.admin!
   end
 
   def grant_access(user)
-    users.delete(user)
-    add(user, as: 'member')
-  end
+    return false unless applicant?(user)
 
-  def add_pending_user(user)
-    add(user, as: 'pending')
+    group_membership_for(user)&.confirmed_member!
   end
 
   def admins
-    User.in_group(self).as(:admin)
+    group_memberships.select(&:admin?).map(&:user)
   end
 
   def confirmed_members
-    User.in_group(self).as(:member) | User.in_group(self).as(:admin)
+    group_memberships.select(&:confirmed_member?).map(&:user)
   end
 
   def members
-    User.in_group(self).as(:member)
+    group_memberships.map(&:user)
```

Comment:
what do you think about something like all_members for clarification?

Reply:
I also thought about `all_members`, but since an applicant is not a member _yet_, I wasn't completely satisfied yet. My thoughts rather led to something like `all_associates`, `all_affiliated`, `all_users`, ...

Reply:
I would prefer `all_users` out of there options, but what do we call the `member?(user)` method then?

Reply:
Okay, then let's take `all_users`. And for the other method, what about `has?(user)`, `contains?(user)`, `user?(user)`? Not perfect yet, but maybe as an idea :)

Reply:
What about just `users`? That way I could just use the `ActiveRecord` default and wouldn't even have to implement a method